### PR TITLE
GH129-Add methods to hand coded NestedElement POCO and add unit tests.

### PR DESCRIPTION
**Pull request by lxatrhea**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/RHEAGROUP/CDP4-SDK-Community-Edition/pulls) open
- [x] I have verified that I am following the CDP4-SDK [code style guidelines](https://raw.githubusercontent.com/RHEAGROUP/CDP4-SDK-Community-Edition/master/.github/CONTRIBUTING.md)
- [x] I have provided test coverage for my change (where applicable)

### Description
Add methods to hand coded NestedElement POCO and add unit tests.
Methods are used in Reporting functionality (DataSource creation).

### CODE REVIEW

File: CDP4Common/Poco/NestedElement.cs [hunk 1]
```diff
@@ -74,5 +74,82 @@ protected DomainOfExpertise GetDerivedOwner()
...
+        /// <returns>
+        /// The <see cref="ElementUsage"/>.
+        /// </returns>
+        public ElementUsage GetElementUsage()
+        {
+            return this.GetElementBase() as ElementUsage;
+        }
+
+        /// <summary>
+        /// Checks if a <see cref="NestedElement"/> contains a specific <see cref="Category"/>.
```
Comment (samatrhea) [inline CDP4Common/Poco/NestedElement.cs:116]: this should be rephrased to:

Asserts whether the NestedElelement is a member of the specific Category

here i am wondering whether super categories are taken into account. There are helper methods for that as well in the SDK, i think extension methods on the Category class

File: CDP4Common/Poco/NestedElement.cs [hunk 2]
```diff
@@ -74,5 +74,82 @@ protected DomainOfExpertise GetDerivedOwner()
...
+        /// </summary>
+        /// <param name="category">
+        /// The <see cref="Category"/>.
+        /// </param>
+        /// <returns>
+        /// True if the <paramref name="category"/> is found, otherwise false.
+        /// </returns>
+        public bool HasCategory(Category category)
```
Comment (samatrhea) [inline CDP4Common/Poco/NestedElement.cs:124]: The name of the method should also be changed. IsMemberOf... then it is semantically right

File: CDP4Common/Poco/NestedElement.cs [hunk 3]
```diff
@@ -74,5 +74,82 @@ protected DomainOfExpertise GetDerivedOwner()
...
+        /// <summary>
+        /// Get the children of a <see cref="NestedElement"/>.
+        /// </summary>
+        /// <param name="nestedElements">
+        /// A list containing all <see cref="NestedElement"/>s.
+        /// </param>
+        /// <returns>
+        /// An <see cref="IEnumerable{NestedElement}"/> containing all children <see cref="NestedElement"/>s.
+        /// </returns>
+        public IEnumerable<NestedElement> GetChildren(List<NestedElement> nestedElements)
```
Comment (samatrhea) [inline CDP4Common/Poco/NestedElement.cs:139]: i think a method like this is already implemented in the NestedElementTreeGenerator. please have a look there.

File: CDP4Common/Poco/NestedElement.cs [hunk 4]
```diff
@@ -74,5 +74,82 @@ protected DomainOfExpertise GetDerivedOwner()
...
+        /// The <see cref="ElementBase"/> representing a <see cref="NestedElement"/>.
+        /// </summary>
+        /// <returns>
+        /// The <see cref="ElementBase"/>.
+        /// </returns>
+        public ElementBase GetElementBase()
+        {
+            return this.IsRootElement
+                ? (ElementBase)this.RootElement
+                : this.ElementUsage.Last();
```
Comment (samatrhea) [inline CDP4Common/Poco/NestedElement.cs:88]: not sure i understand the `Last()` method here? i propose we have a short conversation about that

**Review by samatrhea — CHANGES_REQUESTED** [review]

File: CDP4Common/Poco/NestedElement.cs (hunk 4 above)
Comment (lxatrhea) [reply CDP4Common/Poco/NestedElement.cs:88]: Take a look at the elementUsage property here: [SDK Reference](https://sdk-reference.cdp4.org/#NestedElement)
elementUsage contains an ordered list of hierarchical elementUsages up to the TopElement. So the last ElementUsage will be the correct ElementUsage that the NestedElement represents.
Last() method was already used in the (handcoded POCO) NestedElement.GetDerivedOwner() method.

Comment (samatrhea) [reply CDP4Common/Poco/NestedElement.cs:88]: ok, i sort of remember, thanks for the link to the reference.

File: CDP4Common/Poco/NestedElement.cs (hunk 1 above)
Comment (lxatrhea) [reply CDP4Common/Poco/NestedElement.cs:116]: Yes, they are on the ICategorizableThing. Refactored it.

File: CDP4Common/Poco/NestedElement.cs (hunk 3 above)
Comment (lxatrhea) [reply CDP4Common/Poco/NestedElement.cs:139]: Nothing found there. NestedElementTreeGenerator is only about tree creation and not about reading/traversing it after creation.

**Review by alexatrhea — APPROVED** [review]

**Review by samatrhea — APPROVED** [review]

**Pull request merged**